cylc/flow/tui/util.py

```diff
@@ -373,14 +349,52 @@ def render_node(node, data, type_):
         ret.append(f'{data["name"]}')
         return ret
 
-    if type_ == 'family':
+    if type_ in ['family', 'cycle']:
         return [
             get_task_icon(
                 data['state'],
                 data['isHeld']
             ),
             ' ',
-            data['id'].rsplit('|', 1)[-1]
+            data['id'].rsplit(ID_DELIM, 1)[-1]
         ]
 
-    return data['id'].rsplit('|', 1)[-1]
+    return data['id'].rsplit(ID_DELIM, 1)[-1]
+
+
+PARTS = [
+    'user',
+    'workflow',
+    'cycle_point',
+    'task',
+    'job'
+]
+
+
+def extract_context(selection):
+    """Return a dictionary of all component types in the selection.
+
+    Args:
+        selection (list):
+            List of element id's as extracted from the data store / graphql.
+
+    Examples:
+        >>> extract_context(['a|b', 'a|c'])
+        {'user': ['a'], 'workflow': ['b', 'c']}
+
+        >>> extract_context(['a|b|c|d|e']
+        ... )  # doctest: +NORMALIZE_WHITESPACE
+        {'user': ['a'], 'workflow': ['b'], 'cycle_point': ['c'],
+        'task': ['d'], 'job': ['e']}
+
+    """
```

Comment:
These functions look useful (I feel like I had to write something similar in Cylc UI). Maybe later this will be promoted out of the `tui` module and be used by other code in Cylc Flow (or even Cylc UI Server which imports Cylc Flow).

Reply:
Yes, there is crossover between the two UIs, sadly working with different languages it's difficult to port one to the other.

TUI should always be simple and lightweight, I'd expect the WUI to vastly outstrip it in complexity.